llvm/include/llvm/CodeGen/Analysis.h

```diff
@@ -75,20 +75,23 @@ void ComputeValueVTs(const TargetLowering &TLI, const DataLayout &DL, Type *Ty,
                      SmallVectorImpl<uint64_t> *FixedOffsets,
                      uint64_t StartingOffset);
 
-/// Variant of ComputeValueVTs that also produces the memory VTs.
+/// Variant of ComputeValueVTs that also produces the memory VTs and VT types.
```

Comment:
@arsenm Thanks for the review! If I'm understanding the code correctly, EVTs do contain Type* references, but the EVT created for pointers in ComputeValueVTs is an integer type (`MVT::getIntegerVT(DL.getPointerSizeInBits(AS))`). I don't see a way to easily use the LLVMTy in EVTs for this purpose, but I may very well be mistaken.

It is possible to remove this logic from ComputeValueVTs and have a separate function to flatten the IR type. But there is quite a number of cases where this needs to be done, and the logic is identical to the one in ComputeValueVTs. If you think it preferable to do this, I would be happy to implement it, but I just want to make sure I'm understanding you correctly.

I would also be happy to implement a ComputeValueVTs variant for LLTs if you're going to need that for other things anyway. I don't really know what the GlobalISel transition is though and how involved this would be.

I'm assuming returning booleans (true for pointers) rather than Type*s from ComputeValueVTs does not really make things better for you? This was suggested by phoebewang in the linked issue.

Reply:
@arsenm If you have time, could you still have a look and let me know what approach you'd prefer?

Reply:
Right, the EVTs do unfortunately squash away the pointer types (because they also resolve to integer MVTs, the Type* is dropped). I meant more the other direction, would it work to change this to return the aggregate split Type *s *instead* of the EVT? The client would then have a thin Type*->EVT wrapper

Reply:
@arsenm thank you. I realize this is an old PR – because we were fine working with an LLVM fork, it had less priority to get this merged.

I believe I have addressed your comments. It is possible to return `Type*`s instead of `EVT`s. For this purpose I have added `ComputeValueTypes()`. If you have time I would appreciate advice on what to do with `ComputeValueVTs()`:

1. Remove `ComputeValueVTs()` and rewrite each call site to use `ComputeValueTypes()`. This will be cumbersome as there are many use cases that do not need the extra information and would then need to get the `EVT` from the `Type*`.
2. Do nothing and accept code duplication between `ComputeValueVTs()` and `ComputeValueTypes()`.
3. Rewrite `ComputeValueVTs()` in terms of `ComputeValueTypes()`: iterate over the `Type*`s to get `EVT`s. Reduces duplication, but requires an extra iteration over the result.

I would lean to option 2, but I don't know the general philosophy followed in this code base.